Add location billboards

althea28 opened the pull request:
This PR aims to add billboards to Locations to display the Location Tags present at that location.
<img width="2186" height="1260" alt="image" src="https://github.com/user-attachments/assets/861266d9-1861-4438-8baa-e8e8236084c1" />

When there are no Location Tags present, or the Location only contains Workcell tags, an empty icon will appear instead.
<img width="2243" height="953" alt="image" src="https://github.com/user-attachments/assets/6a8cc82c-023f-4cfb-906b-0d66b82fd3b6" />

**Functionality:**
- Billboards will always rotate to face the camera.
- Billboards show a tooltip of the name of the location tag when the cursor hovers over it.
- The anchor and location can be dragged by clicking and dragging the corresponding billboards.
- Selecting a billboard shows the corresponding Location in the Inspector.


**Changed implementation of the canvas tooltip:**
The existing canvas tooltip implementation causes some of the text to be pushed into the next line. 
<img width="306" height="367" alt="image" src="https://github.com/user-attachments/assets/d4b7a50c-df33-4d8b-b672-f83055399421" />

Hence, part of the canvas tooltips implementation has been modified from this:
```
        let text = self.tips.join("\n");

        egui::Area::new("canvas_tooltip_area".into()).show(ctx, |ui| {
            egui::containers::popup::show_tooltip_text(
                ctx,
                ui.layer_id(),
                "cursor_tooltip".into(),
                text,
            );
        });
```

To this:
```
        let pointer_pos = ctx.input(|i| i.pointer.hover_pos());

        if let Some(pos) = pointer_pos {
            egui::Area::new("canvas_tooltip_area".into())
                .fixed_pos(pos + egui::vec2(16.0, 16.0))
                .show(ctx, |ui| {
                    egui::containers::Frame::popup(ui.style()).show(ui, |ui| {
                        ui.set_max_width(300.0);
                        for tip in &self.tips {
                            ui.label(tip.as_ref());
                        }
                    });
                });
        }
```

From what I have attempted, setting the max_width directly in the egui::containers::popup or egui::Area did not seem to be able to resolve this problem, hence I rewrote the implementation. However, if there is a better way to resolve this, please let me know! From what I can tell, there are no other usages of this Canvas Tooltips, hence as far as I can tell, this new implementation does not appear to affect anything else.

**Known bug:**
When the billboard is selected, the inspector temporarily shows `<Unknown Type> (unsaved)`.

_Steps to recreate the initial bug_:
1. Select a billboard. 
2. In the Inspector, delete the corresponding billboard tag. (e.g. if the Parking billboard was selected, delete the Parking tag).
3. The inspector shows `<Unknown Type> (unsaved)` as the selected billboard was despawned.

https://github.com/user-attachments/assets/ea525f8f-a18f-495c-b5b1-c3f9503d2dc7



The above bug was temporarily resolved by changing the selected object to the corresponding Location under `update_location_for_changed_location_tags`.

```
    for (e, tags, mut billboard_meshes, mut hovered, mut selected) in &mut locations {
        select.write(Select::new(Some(e)));
        …
```

However, there is still a lag when the selected object is changed from the Billboard to the Location. For 1 frame, the `<Unknown Type> (unsaved)` appears again, before the Inspector displays the Location Inspector again. If there is a way to resolve this, please let me know!

https://github.com/user-attachments/assets/3e45f09c-1c9f-4f08-a61e-52f07057b19f

luca-della-vedova commented:
> Known bug:
> When the billboard is selected, the inspector temporarily shows <Unknown Type> (unsaved).
> 
> Steps to recreate the initial bug:
> 
> 1. Select a billboard.
> 2. In the Inspector, delete the corresponding billboard tag. (e.g. if the Parking billboard was selected, delete the Parking tag).
> 3. The inspector shows <Unknown Type> (unsaved) as the selected billboard was despawned.

I feel this is a symptom of something sketchy happening with the selection, and I'd like to check whether the current behavior that we are working around with this workaround + manual system ordering is what you intended.
Specifically, if you click on a billboard the inspector will show the properties of the whole `Location` and not highlight the billboard itself, making me think that we want to tell the user "the location is selected".
However, if the user presses `Delete`, the billboard itself will be deleted, not the location, making me think that we did indeed select the single billboard.
Is this intentional? Do we want to select the billboard or the location? I am fine with both (and I like the idea of deleting billboard through keyboard shortcuts) but:

* If we are selecting the billboard we should highlight it in some way, so users know that if they hit `Delete` it will delete the selected billboard and not the whole location.
* If we are selecting the location we should make sure that clicking `Delete` indeed deletes the whole location instead.

althea28 commented:
> * If we are selecting the location we should make sure that clicking `Delete` indeed deletes the whole location instead.

This is the desired behaviour, my bad for missing this! I've added a few lines within `rmf_site_editor`  >  `src`  >  `keyboard.rs`:
```
            // Check if selected entity has an InspectFor component.
            // Ensures that the deleted entity is also the entity the user sees in the Inspector.
            if let Ok(inspect_for) = inspect_fors.get(selection) {
                delete.write(Delete::new(inspect_for.entity));
            } else {
                delete.write(Delete::new(selection));
            }
```
I added this with the intention to ensure that the deleted entity will also reflect what the user sees in the Inspector, I hope this is an appropriate fix! If there is a more ideal solution for this, I'd be glad to try it out as well.

Aside, in the new commit I've pushed, I've also made a few more edits to:
- Update the location mesh to a Torus instead of a halo, after discussion with Grey on Thursday.
- Fix some other miscellaneous bugs relating to hovering and spawning of the locations.

mxgrey commented on crates/rmf_site_editor/src/keyboard.rs:
```diff
@@ -76,7 +78,13 @@ fn handle_keyboard_input(
         if let Some(selection) = selection.0 {
-            delete.write(Delete::new(selection));
+            // Check if selected entity has an InspectFor component.
+            // Ensures that the deleted entity is also the entity the user sees in the Inspector.
```
I've pushed [f3f8077](https://github.com/open-rmf/rmf_site/pull/372/commits/f3f8077941dbb5f8cf1f270674ea6ad9b3aa4b93) to fix this. Now the handling of `InspectFor` is being done inside the deletion system, so it will work no matter what the source of the Delete event is.

mxgrey reviewed (approved):
I'm very excited about this new billboard feature! Besides the immediate improvements this gives us for Locations, I think we'll see lots of opportunities to expand on this in the future.
